pkg/store/bucket.go

```diff
@@ -2235,45 +2266,63 @@ func newBucketChunkReader(ctx context.Context, block *bucketBlock) *bucketChunkR
 		ctx:      ctx,
 		block:    block,
 		stats:    &queryStats{},
-		preloads: make([][]uint32, len(block.chunkObjs)),
+		preloads: make([][]preloadIdx, len(block.chunkObjs)),
 		chunks:   map[uint64]chunkenc.Chunk{},
 	}
 }
 
-// addPreload adds the chunk with id to the data set that will be fetched on calling preload.
-func (r *bucketChunkReader) addPreload(id uint64) error {
+func (r *bucketChunkReader) Chunk(id uint64) (chunkenc.Chunk, error) {
+	c, ok := r.chunks[id]
+	if !ok {
+		return nil, errors.Errorf("chunk with ID %d not found", id)
+	}
+
+	r.stats.chunksTouched++
+	r.stats.chunksTouchedSizeSum += len(c.Bytes())
+
+	return c, nil
+}
+
+func (r *bucketChunkReader) Close() error {
+	r.block.pendingReaders.Done()
+
+	for _, b := range r.chunkBytes {
+		r.block.chunkPool.Put(b)
+	}
+	return nil
+}
+
+// appPreload adds the chunk with id to the data set that will be fetched on calling preload.
```

Comment:
```suggestion
// addPreload adds the chunk with id to the data set that will be fetched on calling preload.
```

Reply:
Can you update this with `i,j` meanings too?

Reply:
Done. Also, renamed `addPreload()` / `preload()`, as nothing is *pre*loaded anymore.